background.js

```diff
@@ -1,27 +1,171 @@
+        currentTabID = 0
+        chrome.tabs.query({}, function (tabs) {
+           chrome.tabs.query({active: true, currentWindow: true}, function (tabs) {
+               const activeTab = tabs[0]
+               updateTabId(activeTab.id)
+               if (activeTab.url.includes("developer.apple.com")) {
+                   updateLogo(true)
+               } else {
+                   updateLogo(false)
+               }
+           })
+
+            refreshRequested = true
+
+            let running = []
+            for (let tab of tabs) {
+                if (tab.url.includes("developer.apple.com")) {
+                    running.push(chrome.tabs.reload(tab.id))
+                }
+            }
+
+            Promise.allSettled(running).then()
         });
-    } else if (request.type === initalRequestMethod) {
+        return true
+    } else if (request.type === initialRequestMethod) {
         sendResponse({shouldTranslate: shouldTranslate});
+    } else if (request.type === translatedRequestMethod) {
+        currentTranslatedPage.push(sender.tab.id)
+        updateLogo(true)
+        return true
     }
 });
 
 function delay(i) {
   setTimeout(() => {
-    console.log(i);
   }, i * 1000);
 }
+
+chrome.tabs.onUpdated.addListener(function(tabId, info, tab) {
+    if (refreshRequested) {
+        refreshRequested = false
+        return true
+    }
+
+    chrome.tabs.query({ active: true, currentWindow: true }, function (tabs) {
+        const activeTab = tabs[0]
+        updateTabId(activeTab.id)
+        currentTranslatedPage = removeItemAll(currentTranslatedPage, previousTabID)
+        if (activeTab.url.includes("developer.apple.com")) {
+            updateLogo(true)
+        } else {
+            updateLogo(false)
+        }
+
+        chrome.tabs.sendMessage( tabId, {
+            message: pageSwitchedRequestMethod,
+            url: activeTab.url
+        }).then()
+    });
+
+    return true
+});
+
+chrome.tabs.onActivated.addListener(function(activeInfo) {
+    chrome.tabs.get(activeInfo.tabId, function(tab) {
+        updateTabId(tab.id)
+        if (tab.url && !(tab.url.includes('developer.apple.com/'))) {
+            updateLogo(false)
+            currentTranslatedPage = removeItemAll(currentTranslatedPage, currentTabID)
+        } else if (tab.url && (tab.url.includes('developer.apple.com/'))) {
+            updateLogo(true)
+        }
+    });
+});
+
+function removeItemAll(arr, value) {
+    var index = arr.indexOf(value);
+    if (index > -1) {
+        arr.splice(index, 1);
+    }
+    return arr;
+}
+function updateTabId(id) {
+    previousTabID = currentTabID
+    currentTabID = id
+}
+
+function updateLogo(active) {
+    if (active) {
+        if (shouldTranslate) {
+            if (currentTranslatedPage.includes(currentTabID)) {
+                setIcon("/source/intro/swiftLogo-translating.png")
+            } else {
+                setIcon("/source/intro/swiftLogo-running.png")
```

Comment:
running status is so fast, I think if we really want to identify it, we can make a notable loading status.

Reply:
Running status, the orange one, means that the user is currently viewing an website that SwiftGG will work on but this page isn’t supported. And green means the user is viewing a supported and translated page.

Reply:
It doesn’t mean running.

Reply:
sure, I think users may can't get the real meanning. lol

Reply:
Orange just means run. And green means translated.

Reply:
It doesn't mean running but means run

Reply:
Orange means SwiftGG works on this website and green means SwiftGG can translate this page.

Reply:
If we support other websites in the future, such as hacking with swift and others. The user only needs to see whether there is an indicator.

Reply:
## Tip on the Extension Icon

The extension Icon is an excellent place to tell users the plugin's status.

* When the translating option is turned off, the icon will become grey.

![CleanShot 2023-07-15 at 23 01 07@2x](https://github.com/SwiftGGTeam/swiftgg-trans-plugin/assets/46642225/b47c14ad-a027-40ab-abec-2445853d150e)

* When the translating option is turned on, the user is on a non-related website, like `google.com`, the logo will become a normal one.

![CleanShot 2023-07-15 at 23 02 06@2x](https://github.com/SwiftGGTeam/swiftgg-trans-plugin/assets/46642225/5ea3b122-edda-4dd0-a382-b3822560875c)

* When the translating option is turned on, the user is on a related website, like `developer.apple.com`,  but the page is not supported; the logo will become a normal one with an orange indicator. Example: `https://developer.apple.com/tutorials/sample-apps`

![CleanShot 2023-07-15 at 23 03 25@2x](https://github.com/SwiftGGTeam/swiftgg-trans-plugin/assets/46642225/3ac83a17-8282-4676-b905-417f7320c379)

* When the translating option is turned on, the user is on a related website, like `developer.apple.com`, and the page is supported; the logo will become a normal one with a green indicator. Example: `https://developer.apple.com/tutorials/swiftui/creating-and-combining-views`

![CleanShot 2023-07-15 at 23 04 04@2x](https://github.com/SwiftGGTeam/swiftgg-trans-plugin/assets/46642225/61ddf845-735d-4c27-a81b-10fd1ea04d0a)

This function works well throughout the whole process and can detect tab updating, switching and closing events.